### ISSUE TITLE
systemd-networkd: PD doesn't work in some case

### ISSUE THREAD

**Issue by LetzteInstanz** [systemd/systemd#22068]
Hello. The prefix delegation feature is broken for me since 250v.

The configuration, that works in 249v:
`10-wan.network`:
```
[Match]
Name=wan

[Network]
IgnoreCarrierLoss=true
KeepConfiguration=dhcp-on-stop
DHCP=yes
IPForward=yes
IPv6AcceptRA=true
IPv6LinkLocalAddressGenerationMode=random
IPv6Token=prefixstable
DNS=2dc0:a17b:7415:0000:021b:21ff:fe63:2216

[DHCPv4]
UseDNS=false

[DHCPv6]
ForceDHCPv6PDOtherInformation=true
WithoutRA=information-request
UseDNS=false

[IPv6AcceptRA]
UseDNS=false
```
`11-lan.network`:
```
[Match]
Name=lan

[Network]
IgnoreCarrierLoss=true
KeepConfiguration=dhcp-on-stop
IPForward=yes
IPv6SendRA=false
DHCPv6PrefixDelegation=true
Address=192.168.153.1/24
DNS=2dc0:a17b:7415:0000:021b:21ff:fe63:2216
Domains=lan.rnet.ru

[Address]
Address=2dc0:a17b:7415:0000:021b:21ff:fe63:2216/64

[Neighbor]
Address=192.168.153.254
LinkLayerAddress=ff:ff:ff:ff:ff:ff

[IPv6SendRA]
RouterLifetimeSec=2592000
DNS=2dc0:a17b:7415:0:21b:21ff:fe63:2216
Domains=lan.rnet.ru
DNSLifetimeSec=7776000

[IPv6Prefix]
Prefix=2dc0:a17b:7415:0000::/64
PreferredLifetimeSec=2592000
ValidLifetimeSec=7776000
Assign=false

[DHCPPrefixDelegation]
UplinkInterface=wan
```
The configuration, that must work in 250v in my opinion:
`10-wan.network`:
```
[Match]
Name=wan

[Network]
IgnoreCarrierLoss=true
KeepConfiguration=dhcp-on-stop
DHCP=yes
IPForward=yes
IPv6AcceptRA=true
IPv6LinkLocalAddressGenerationMode=random
IPv6Token=prefixstable
DNS=2dc0:a17b:7415:0000:021b:21ff:fe63:2216

[DHCPv4]
UseDNS=false

[DHCPv6]
WithoutRA=solicit
UseDNS=false

[IPv6AcceptRA]
UseDNS=false
```
`11-lan.network`:
```
[Match]
Name=lan

[Network]
IgnoreCarrierLoss=true
KeepConfiguration=dhcp-on-stop
IPForward=yes
IPv6SendRA=false
DHCPPrefixDelegation=true
Address=192.168.153.1/24
DNS=2dc0:a17b:7415:0000:021b:21ff:fe63:2216
Domains=lan.rnet.ru

[Address]
Address=2dc0:a17b:7415:0000:021b:21ff:fe63:2216/64

[Neighbor]
Address=192.168.153.254
LinkLayerAddress=ff:ff:ff:ff:ff:ff

[IPv6SendRA]
RouterLifetimeSec=2592000
DNS=2dc0:a17b:7415:0:21b:21ff:fe63:2216
Domains=lan.rnet.ru
DNSLifetimeSec=7776000

[IPv6Prefix]
Prefix=2dc0:a17b:7415:0000::/64
PreferredLifetimeSec=2592000
ValidLifetimeSec=7776000
Assign=false

[DHCPPrefixDelegation]
UplinkInterface=wan
```
My ISP provides me SLAAC through NDP and uses a DHCP-server for only prefix delegation and providing other configuration information (DNS servers, DNS search list).

I dumped DHCP request messages of the both versions of systemd and compared them.

249v sends a solicit message with the `IA_PD` option and without the `IA_NA` option, and the DHCP server responds an advertise message with a prefix. 250v sends a solicit message with the both options: `IA_PD` and `IA_NA`, and the DHCP server responds an advertise message with the status code `NoAddrsAvail`, because the DHCP server doesn't provide adresses.

I took a dump of the original solicit message from 250v, removed the `IA_NA` option from the message, sent it by `tcpreplay`, and the DHCP server responded a message with the prefix.

When I send the original solicit message:
`# tcpdump -e -xx -n -vvv -i wan 'ip6 and udp and (port 546 or port 547)'`:
```
01:12:41.826271 70:85:c2:25:bc:62 > 33:33:00:01:00:02, ethertype IPv6 (0x86dd), length 150: (flowlabel 0x679cb, hlim 1, next-header UDP (17) payload length: 96) fe80::759c:aee9:17d1:fb7d.546 > ff02::1:2.547: [udp sum ok] dhcp6 solicit (xid=f16e8d (rapid-commit) (IA_NA IAID:1719374634 T1:0 T2:0) (Client-FQDN) (IA_PD IAID:1719374634 T1:0 T2:0) (option-request DNS-server DNS-search-list NTP-server SNTP-servers) (client-ID vid 0000ab1107b7d202) (elapsed-time 0))
        0x0000:  3333 0001 0002 7085 c225 bc62 86dd 6006
        0x0010:  79cb 0060 1101 fe80 0000 0000 0000 759c
        0x0020:  aee9 17d1 fb7d ff02 0000 0000 0000 0000
        0x0030:  0000 0001 0002 0222 0223 0060 9300 01f1
        0x0040:  6e8d 000e 0000 0003 000c 667b 932a 0000
        0x0050:  0000 0000 0000 0027 0008 0106 7365 7276
        0x0060:  6572 0019 000c 667b 932a 0000 0000 0000
        0x0070:  0000 0006 0008 0017 0018 0038 001f 0001
        0x0080:  000e 0002 0000 ab11 07b7 d202 fe79 fe08
        0x0090:  0008 0002 0000
01:12:41.829640 00:04:4e:87:f0:00 > 70:85:c2:25:bc:62, ethertype IPv6 (0x86dd), length 117: (class 0xe0, hlim 255, next-header UDP (17) payload length: 63) fe80::204:4eff:fe87:f000.547 > fe80::759c:aee9:17d1:fb7d.546: [udp sum ok] dhcp6 advertise (xid=f16e8d (server-ID hwaddr type 1 00044e87f000) (client-ID vid 0000ab1107b7d202) (status-code NoAddrsAvail))
        0x0000:  7085 c225 bc62 0004 4e87 f000 86dd 6e00
        0x0010:  0000 003f 11ff fe80 0000 0000 0000 0204
        0x0020:  4eff fe87 f000 fe80 0000 0000 0000 759c
        0x0030:  aee9 17d1 fb7d 0223 0222 003f 5f7d 02f1
        0x0040:  6e8d 0002 000a 0003 0001 0004 4e87 f000
        0x0050:  0001 000e 0002 0000 ab11 07b7 d202 fe79
        0x0060:  fe08 000d 000f 0002 4e4f 4144 4452 532d
        0x0070:  4156 4149 4c
```
When I send the message without the `IA_NA` option:
```
01:15:24.099210 70:85:c2:25:bc:62 > 33:33:00:01:00:02, ethertype IPv6 (0x86dd), length 134: (flowlabel 0x679cb, hlim 1, next-header UDP (17) payload length: 80) fe80::759c:aee9:17d1:fb7d.546 > ff02::1:2.547: [udp sum ok] dhcp6 solicit (xid=f16e8d (rapid-commit) (Client-FQDN) (IA_PD IAID:1719374634 T1:0 T2:0) (option-request DNS-server DNS-search-list NTP-server SNTP-servers) (client-ID vid 0000ab1107b7d202) (elapsed-time 0))
        0x0000:  3333 0001 0002 7085 c225 bc62 86dd 6006
        0x0010:  79cb 0050 1101 fe80 0000 0000 0000 759c
        0x0020:  aee9 17d1 fb7d ff02 0000 0000 0000 0000
        0x0030:  0000 0001 0002 0222 0223 0050 8cd5 01f1
        0x0040:  6e8d 000e 0000 0027 0008 0106 7365 7276
        0x0050:  6572 0019 000c 667b 932a 0000 0000 0000
        0x0060:  0000 0006 0008 0017 0018 0038 001f 0001
        0x0070:  000e 0002 0000 ab11 07b7 d202 fe79 fe08
        0x0080:  0008 0002 0000
01:15:24.100048 00:04:4e:87:f0:00 > 70:85:c2:25:bc:62, ethertype IPv6 (0x86dd), length 192: (class 0xe0, hlim 255, next-header UDP (17) payload length: 138) fe80::204:4eff:fe87:f000.547 > fe80::759c:aee9:17d1:fb7d.546: [udp sum ok] dhcp6 advertise (xid=f16e8d (server-ID hwaddr type 1 00044e87f000) (client-ID vid 0000ab1107b7d202) (DNS-server 2a00:1838:10:2:3426:89ff:fe78:c2f3 2a00:1838:10:2:7c52:d5ff:fe4d:cb4) (DNS-search-list rnet.ru.) (IA_PD IAID:1719374634 T1:300 T2:480 (IA_PD-prefix 2a00:1838:bbbb:1200::/56 pltime:600 vltime:1800)))
        0x0000:  7085 c225 bc62 0004 4e87 f000 86dd 6e00
        0x0010:  0000 008a 11ff fe80 0000 0000 0000 0204
        0x0020:  4eff fe87 f000 fe80 0000 0000 0000 759c
        0x0030:  aee9 17d1 fb7d 0223 0222 008a 0263 02f1
        0x0040:  6e8d 0002 000a 0003 0001 0004 4e87 f000
        0x0050:  0001 000e 0002 0000 ab11 07b7 d202 fe79
        0x0060:  fe08 0017 0020 2a00 1838 0010 0002 3426
        0x0070:  89ff fe78 c2f3 2a00 1838 0010 0002 7c52
        0x0080:  d5ff fe4d 0cb4 0018 0009 0472 6e65 7402
        0x0090:  7275 0000 1900 2966 7b93 2a00 0001 2c00
        0x00a0:  0001 e000 1a00 1900 0002 5800 0007 0838
        0x00b0:  2a00 1838 bbbb 1200 0000 0000 0000 0000
```
Maybe introduce `RequestOnlyPrefix=` option into the section `[DHCPv6]`? Or `Request=address`/`prefix`/`both`?

**Comment by bluca** [comment]
Please attach the networkd log (debug level)

**Comment by LetzteInstanz** [comment]
> Please attach the networkd log (debug level)

[systemd-networkd.log](https://github.com/systemd/systemd/files/7837882/systemd-networkd.log)

**Comment by yuwata** [comment]
From the provided log:
```
wan: DHCPv6 client: Received Advertise message with non-zero status: NOADDRS-AVAIL: No addresses available
wan: DHCPv6 client: Failed to process received advertise message, ignoring: Invalid argument
```

**Comment by yuwata** [comment]
Fix is waiting in #22080. If possible, could you test the PR? Thank you for the report.

**Comment by LetzteInstanz** [comment]
> Fix is waiting in #22080. If possible, could you test the PR? Thank you for the report.

Unfortunately, no. Thank you for the fix.

**Comment by yuwata** [comment]
@LetzteInstanz I noticed we already had UseAddress= setting. Let's use it to control what the client requests. Fix is waiting in #22083.

**Comment by LetzteInstanz** [comment]
> @LetzteInstanz I noticed we already had UseAddress= setting. Let's use it to control what the client requests. Fix is waiting in #22083.

Thanks! I have understood all now.
